lib/ui/painting.dart

```diff
@@ -3906,27 +3906,65 @@ class FragmentProgram extends NativeFieldWrapperClass1 {
   /// [A current specification of valid SPIR-V is here.](https://github.com/flutter/engine/blob/master/lib/spirv/README.md)
   /// SPIR-V not meeting this specification will throw an exception.
   static Future<FragmentProgram> compile({
```

Comment:
I don't think we should combine this new functionality into the old API. Instead we should start from scratch with a new one and deprecate this.

maybe we'd always want to retain an API for loading bytes from the dart heap - just to support certain dynamic cases, but for bundled assets ideally we would only pass a reference to the asset (probably the string key) and let the engine do the loading.

That would also allow me to support hot reload with no framework side changes.

Also, can impellerc include uniform counts into the blob it creates so users don't need to manage that?

Reply:
Yeah all of that makes sense to me. Not sure what new API here would make life easiest. Do you have any suggestions?

Reply:
Something like:

```
FragmentProgram.fromAsset(String key)

```
or if needs to be async, then

```
static Future<FragmentProgram> fromAsset(String key)
```

Depending on the ownership needs of the engine, we could also use immutable buffer but that implies at least one copy.

Where the key would be an asset containing platform specific (metal, sksl, et cetera) blob that had enough information.

That said, we also don't need to come up with this API in this PR, and could instead leave the old API for now and land the compiler changes to allow for experimentation?